src/domain/account.ts

```diff
@@ -1,264 +1,342 @@
-        network
-      ),
-    getNotRestoredWatchIdentity: (network: NetworkString) =>
-      newMasterPublicKey(data.masterXPub, data.masterBlindingKey, network),
-    getDeepRestorer: (network: NetworkString) => {
-      const pubkey = newMasterPublicKey(data.masterXPub, data.masterBlindingKey, network);
-      return makeRestorerFromChainAPI<MasterPublicKey>(
-        pubkey,
-        (isChange, index) => pubkey.getAddress(isChange, index).address.confidentialAddress
+// Account is a readonly way to interact with the account data (transactions, utxos, scripts, etc.)
+export class Account {
+  private chainSource: ChainSource;
+  private network: networks.Network;
+  private node: BIP32Interface;
+  private blindingKeyNode: Slip77Interface;
+  private walletRepository: WalletRepository;
+  readonly name: string;
+
+  static BASE_DERIVATION_PATH = "m/84'/1776'/0'";
+  static BASE_DERIVATION_PATH_LEGACY = "m/84'/0'/0'";
+  static BASE_DERIVATION_PATH_TESTNET = "m/84'/1'/0'";
+
+  constructor({
+    chainSource,
+    masterBlindingKey,
+    masterPublicKey,
+    network,
+    walletRepository,
+    name,
+  }: AccountOpts) {
+    this.name = name;
+    this.chainSource = chainSource;
+    this.network = networks[network];
+    this.node = bip32.fromBase58(masterPublicKey);
+    this.blindingKeyNode = slip77.fromMasterBlindingKey(masterBlindingKey);
+    this.walletRepository = walletRepository;
+  }
+
+  deriveBlindingKey(script: Buffer): { publicKey: Buffer; privateKey: Buffer } {
+    if (!this.blindingKeyNode)
+      throw new Error('No blinding key node, Account cannot derive blinding key');
+    const derived = this.blindingKeyNode.derive(script);
+    if (!derived.publicKey || !derived.privateKey) throw new Error('Could not derive blinding key');
+    return { publicKey: derived.publicKey, privateKey: derived.privateKey };
+  }
+
+  // Derive a range from start to end index of public keys applying the base derivation path
+  async deriveBatch(
+    start: number,
+    end: number,
+    isInternal: boolean,
+    updateCache = true
+  ): Promise<Buffer[]> {
+    if (!this.node) throw new Error('No node, Account cannot derive scripts');
+    const chain = isInternal ? 1 : 0;
+    const scripts = [];
+    for (let i = start; i < end; i++) {
+      const child = this.node.derive(chain).derive(i);
+      const p2wpkh = payments.p2wpkh({ pubkey: child.publicKey, network: this.network });
+      const script = p2wpkh.output;
+      if (!script) continue;
+      scripts.push(script);
+    }
+
+    if (this.walletRepository && updateCache) {
+      // persist the script details in the wallet repository
+      await this.walletRepository.updateScriptDetails(
+        Object.fromEntries(
+          scripts.map((script) => [
+            script.toString('hex'),
+            {
+              accountName: this.name,
+              network: this.network.name as NetworkString,
+              derivationPath: `m/${chain}/${start}`,
```

Comment:
Why this derivation path is missing the "coin" segment 84' or 1776'?
Why "chain" is missing the ' ?
Why the index is always `start`?

Reply:
> Why this derivation path is missing the "coin" segment 84' or 1776'?

The "coin" section is a part of the _hardened_ path and `Account` class does not have access to any private keys (so unable to derive hardened indexes). The coin section and all others hardened ones are derived at the account creation step and pass to `Account` class via the account masterPublicKey (xpub).

> Why "chain" is missing the ' ? 

As far as I know, chain should not be hardened.

> Why the index is always `start`?

nice catch, fixed, now is `start + index`.